### ISSUE TITLE
ScmVersion time sould be UTC for (dirty) repo with local_scheme 'node-and-date' or 'node-and-timestamp'

### ISSUE THREAD

**Issue by con-f-use** [pypa/setuptools-scm#391]
I noticed that  for a repository (maybe only if it's dirty), there is no time zone information passed to the `local_scheme` of choice and the time information is in local time:

```python
from setuptools import setup

def loc_vers(*a, **kw):
    scm = a[0]
    print(scm.time)  # 2020-01-14 14:56:16.031252 (my local time, UTC would be an hour earlier)
    print(scm.time.tzname())  # None
    return ""

setup(
    use_scm_version=dict(local_scheme=loc_vers),
    # ...
)
```

Seeing that (linux epoch) timestamps are supposed to be UTC and git repository timestams are UTC as well, I think this should be UTC or at least have a `tzinfo` set.

**Comment by con-f-use** [comment]
I've been doing a little thinking on this one. For the normal `ScmVersion.time`, we should use datetime.datetime.now(datetime.timezone.utc)` instead of just `datetime.datetime()`. That would include timezone information make the time location independent. Since it's part of the version and version should hopefully be location independent, that makes sense to me.

Also, I'd like to add a new attribute to the `ScmVersion` class that tracks the "commit time", e.g. for git that would come from `git show -s --format=%ct` which is a UNIX timestamp. The thinking here is, that `node-and-date` and `node-and-timestamp` local schemes should use:

 * The system's local time when the commit is dirty
 * The time of commit, when the commit is not dirty

Because currently, we get a new version (i.e. a higher timestamp) even if a package is build a second time from the same unchanged commit.

In any case, all times should have timezone information or be in UTC.

How does that sound conceptually?

@RonnyPfannschmidt  at some point I'd love a talk with if you're still in Innsbruck and the whole virus-thing is over.

**Comment by RonnyPfannschmidt** [comment]
i like the general direction, the exact details will be a bit more tricky,
we may be ok to use `.utcnow()` instead of now

as for having a talk, i strongly propose a virtual one as we don't know what the future may bring

**Comment by con-f-use** [comment]
Agreed, I was planning to discuss this issue further here. I think a nice solution is `datetime.datetime.now(datetime.timezone.utc)` because the date and time are part of the version and version should be (ideally) location independent :wink: 

Sadly `.utcnow()` does not set `tzinfo`:

```
>>> import datetime
>>> datetime.datetime.now(datetime.timezone.utc) 
datetime.datetime(2020, 3, 18, 16, 5, 21, 883066, tzinfo=datetime.timezone.utc)
>>> datetime.datetime.utcnow()                                                                                                                
datetime.datetime(2020, 3, 18, 16, 10, 10, 820451)
```

The in-person talk, I'd like to have because of other general ideas for python packaging and your relative closeness. But that can wait almost indefinitely.

**Comment by RonnyPfannschmidt** [comment]
also as a note, with the daycare shudown and the toddler at home, im not sure how much room i will have for other things in near future

**Comment by con-f-use** [comment]
Understood. Strange times we live in.

I'll probably make two pull-requests, one for the `datetime.datetime.now(datetime.timezone.utc)` thing and one for the `commit_time` thing. The latter is the more controversial and can be discussed there and then. That is if you are okay with that. If a `commit_time` PR has no chance of being considered, I'll not waste my time. I realize both are very low on the priority list at the moment and I'm fine with waiting, just want to avoid wasting work. :grin:

**Comment by con-f-use** [comment]
Turns out `datetime.timezone.utc` is Python 3 only. 

So it's either a thrid-party library like `pytz` or `python-dateutil` like this:

```
import datetime, pytz
datetime.datetime.now(pytz.utc)
``` 

or:

```
import datetime

_ZERO = datetime.timedelta(0)

class UTC(datetime.tzinfo):
    """UTC"""
    def utcoffset(self, dt): return _ZERO
    def tzname(self, dt): return "UTC"
    def dst(self, dt): return _ZERO

utc = UTC()

datetime.datetime.now(utc)
```

**Comment by RonnyPfannschmidt** [comment]
closed by #408
